lib/SILOptimizer/Utils/GenericCloner.cpp

```diff
@@ -73,6 +73,10 @@ void GenericCloner::populateCloned() {
     SILArgument *OrigArg = *I;
     RegularLocation Loc((Decl *)OrigArg->getDecl());
     AllocStackInst *ASI = nullptr;
+    // TODO: Special handling of archetypes with have constraint layouts.
```

Comment:
I only added a comment here, no code changes. And yes, those archetypes will be not be passed indirectly.

IIRC, the logic here is something like:
- the caller used the a generic apply and was passing the parameters indirectly.
- All the code in the non-specialized callee's body is assuming that parameters are passed indirectly.
- then we try to specialize the callee and detect that we can make some parameters concrete.
- when we clone the callee's body during specialization, we rewrite it in such a way that it knows which parameters are not indirect anymore. 
- The way how it is currently modeled is to alloc_stack them in the entry block, which allows for very easy cloning, since all the code which was assuming indirect access doesn't need to be changed. Later on most of this gets promoted into direct access anyways.
- In principle, we could have rewritten indirect access into direct access, but this is a separate issue.

@eeckstein may know more about the reason why we don't rewrite indirect into direct access currently, because he implemented this logic.